File: packages/python/codex/src/anosys_sdk_codex/installer.py
"""
Safely manage ~/.codex/config.toml with AnoSys-managed notify hook.

Handles idempotent install/uninstall of the AnoSys Codex notify hook
and ~/.codex/anosys-env.sh configuration.
"""

from __future__ import annotations

import json
import os
import re
import shutil
import sys
import tempfile
import time
from pathlib import Path
from typing import Any, Dict, List, Optional

try:
    import tomllib  # Python 3.11+
except ImportError:
    try:
        import tomli as tomllib  # type: ignore[no-redef]
    except ImportError:
        tomllib = None  # type: ignore[assignment]


HOOK_COMMAND = "anosys-codex run"
INGESTION_URL = "https://api.anosys.ai/ingestion"


def get_codex_home() -> Path:
    """Return the active Codex home directory, respecting CODEX_HOME."""
    custom = os.environ.get("CODEX_HOME", "").strip()
    if custom:
        return Path(os.path.expandvars(os.path.expanduser(custom))).resolve()
    return Path.home() / ".codex"


def get_config_path() -> Path:
    return get_codex_home() / "config.toml"


def get_env_path() -> Path:
    return get_codex_home() / "anosys-env.sh"


def backup(path: Path | None = None) -> Optional[Path]:
    """Create a timestamped backup of the config file."""
    if path is None:
        path = get_config_path()
    if path.is_file():
        timestamp = time.strftime("%Y%m%d%H%M%S", time.gmtime())
        backup_path = path.with_name(f"{path.name}.{timestamp}.bak")
        shutil.copy2(path, backup_path)
        return backup_path
    return None


def load_toml(path: Path) -> Dict[str, Any]:
    """Read and parse TOML file. Returns empty dict if file does not exist."""
    if not path.is_file():
        return {}
    try:
        content = path.read_text(encoding="utf-8")
        if not content.strip():
            return {}
        if tomllib is not None:
            return tomllib.loads(content)
        return _fallback_parse_toml(content)
    except Exception as exc:
        sys.exit(
            f"ERROR: {path} contains invalid TOML and cannot be parsed safely.\n"
            f"  {exc}\n"
            f"Please fix the file manually and re-run the installer."
        )


def _fallback_parse_toml(content: str) -> Dict[str, Any]:
    res: Dict[str, Any] = {}
    notify_match = re.search(r'^\s*notify\s*=\s*\[(.*?)\]', content, re.MULTILINE | re.DOTALL)
    if notify_match:
        items = re.findall(r'["\']([^"\']+)["\']', notify_match.group(1))
        res["notify"] = items
    return res


def _serialize_toml_value(val: Any) -> str:
    if isinstance(val, bool):
        return "true" if val else "false"
    if isinstance(val, (int, float)):
        return str(val)
    if isinstance(val, str):
        escaped = val.replace("\\", "\\\\").replace('"', '\"')
        return f'"{escaped}"'
    if isinstance(val, list):
        items_str = ", ".join(_serialize_toml_value(x) for x in val)
        return f"[{items_str}]"
    if isinstance(val, dict):
        items_str = ", ".join(f"{k} = {_serialize_toml_value(v)}" for k, v in val.items())
        return f"{{ {items_str} }}"
    return f'"{str(val)}"'


def write_toml_atomic(path: Path, data: Dict[str, Any]) -> None:
    """Safely write TOML data atomically."""
    path.parent.mkdir(parents=True, exist_ok=True)
    lines: List[str] = []

    for k, v in data.items():
        if not isinstance(v, dict):
            lines.append(f"{k} = {_serialize_toml_value(v)}")

    for k, v in data.items():
        if isinstance(v, dict):
            lines.append("")
            lines.append(f"[{k}]")
            for sub_k, sub_v in v.items():
                lines.append(f"{sub_k} = {_serialize_toml_value(sub_v)}")

    content = "\n".join(lines).strip() + "\n"
    fd, tmp_path = tempfile.mkstemp(dir=path.parent, prefix="config_", suffix=".tmp")
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            fh.write(content)
        shutil.move(tmp_path, path)
    except Exception:
        try:
            os.unlink(tmp_path)
        except OSError:
            pass
        raise


def has_anosys_hook(data: Dict[str, Any]) -> bool:
    notify_list = data.get("notify", [])
    if isinstance(notify_list, str):
        notify_list = [notify_list]
    if isinstance(notify_list, list):
        return any("anosys-codex" in str(x) for x in notify_list)
    return False


def update_codex_config(hook_command: str = HOOK_COMMAND, path: Path | None = None) -> bool:
    if path is None:
        path = get_config_path()

    data = load_toml(path)
    notify_list = data.get("notify", [])
    if isinstance(notify_list, str):
        notify_list = [notify_list]
    elif not isinstance(notify_list, list):
        notify_list = []

    existing = [x for x in notify_list if "anosys-codex" in str(x)]
    if existing and hook_command in notify_list:
        return False

    new_notify = [x for x in notify_list if "anosys-codex" not in str(x)]
    new_notify.append(hook_command)
    data["notify"] = new_notify

    write_toml_atomic(path, data)
    return True


def remove_codex_config(path: Path | None = None) -> bool:
    if path is None:
        path = get_config_path()
    if not path.is_file():
        return False

    data = load_toml(path)
    notify_list = data.get("notify", [])
    if isinstance(notify_list, str):
        notify_list = [notify_list]
    elif not isinstance(notify_list, list):
        return False

    new_notify = [x for x in notify_list if "anosys-codex" not in str(x)]
    if len(new_notify) == len(notify_list):
        return False

    if new_notify:
        data["notify"] = new_notify
    else:
        data.pop("notify", None)

    write_toml_atomic(path, data)
    return True


def update_codex_env(
    api_key: str = "",
    redaction: bool = False,
    endpoint_url: str = INGESTION_URL,
    path: Path | None = None,
) -> None:
    if path is None:
        path = get_env_path()

    path.parent.mkdir(parents=True, exist_ok=True)
    lines = [
        "# AnoSys Codex CLI configuration",
        f'export ANOSYS_HOOK_ENDPOINT_URL="{endpoint_url}"',
    ]
    if api_key:
        lines.append(f'export ANOSYS_HOOK_APIKEY="{api_key}"')
    if redaction:
        lines.append('export REDACTION="true"')
    else:
        lines.append('export REDACTION="false"')

    content = "\n".join(lines) + "\n"
    path.write_text(content, encoding="utf-8")
    try:
        os.chmod(path, 0o600)
    except OSError:
        pass


def remove_codex_env(path: Path | None = None) -> bool:
    if path is None:
        path = get_env_path()
    if path.is_file():
        try:
            path.unlink()
            return True
        except OSError:
            return False
    return False


def validate_api_key(api_key: str, key_type: str = "cc") -> bool:
    if not api_key:
        return False
    try:
        import urllib.parse
        import urllib.request

        url = f"https://console.anosys.ai/api/resolveapikeys?apikey={urllib.parse.quote(api_key)}"
        req = urllib.request.Request(
            url, headers={"User-Agent": "anosys-codex-installer"}
        )
        with urllib.request.urlopen(req, timeout=10) as response:
            if response.status < 200 or response.status >= 300:
                return False
            data = json.loads(response.read().decode("utf-8"))
            api_url = data.get("apiUrl")
            if not api_url:
                return False
            lower = str(key_type).lower()
            if lower in ("cc", "claudecode", "codex"):
                return "/cc/" in api_url
            if lower in ("t", "otel"):
                return "/t/" in api_url
            return False
    except Exception:
        return False

File: packages/js/codex/src/installer.js
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const https = require('https');

const HOOK_COMMAND = 'anosys-codex run';
const INGESTION_URL = 'https://api.anosys.ai/ingestion';

function getCodexHome() {
  const custom = (process.env.CODEX_HOME || '').trim();
  if (custom) {
    return path.resolve(custom.replace(/^~(?=$|\/|\\)/, os.homedir()));
  }
  return path.join(os.homedir(), '.codex');
}

function getConfigPath() {
  return path.join(getCodexHome(), 'config.toml');
}

function getEnvPath() {
  return path.join(getCodexHome(), 'anosys-env.sh');
}

function backup(customPath) {
  const filePath = customPath || getConfigPath();
  if (fs.existsSync(filePath)) {
    const timestamp = new Date().toISOString().replace(/[-:T.]/g, '').slice(0, 14);
    const backupPath = `${filePath}.${timestamp}.bak`;
    fs.copyFileSync(filePath, backupPath);
    return backupPath;
  }
  return null;
}

function loadToml(filePath) {
  if (!fs.existsSync(filePath)) return {};
  try {
    const content = fs.readFileSync(filePath, 'utf8');
    const result = {};
    const notifyMatch = content.match(/^\s*notify\s*=\s*\[(.*?)\]/ms);
    if (notifyMatch) {
      const items = [];
      const regex = /["']([^"']+)["']/g;
      let match;
      while ((match = regex.exec(notifyMatch[1])) !== null) {
        items.push(match[1]);
      }
      result.notify = items;
    }
    return result;
  } catch (err) {
    console.error(`ERROR: Failed to parse TOML at ${filePath}:`, err.message);
    return {};
  }
}

function writeTomlAtomic(filePath, data) {
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });

  let lines = [];
  if (Array.isArray(data.notify)) {
    const serialized = data.notify.map(x => `"${x.replace(/"/g, '\\"')}"`).join(', ');
    lines.push(`notify = [${serialized}]`);
  }

  // Preserve or write other sections if any
  for (const [key, value] of Object.entries(data)) {
    if (key === 'notify') continue;
    if (typeof value === 'object' && !Array.isArray(value) && value !== null) {
      lines.push(`\n[${key}]`);
      for (const [subKey, subVal] of Object.entries(value)) {
        lines.push(`${subKey} = "${String(subVal).replace(/"/g, '\\"')}"`);
      }
    } else {
      lines.push(`${key} = "${String(value).replace(/"/g, '\\"')}"`);
    }
  }

  const tmpPath = path.join(dir, `.tmp_${Date.now()}_config.toml`);
  fs.writeFileSync(tmpPath, lines.join('\n') + '\n', 'utf8');
  fs.renameSync(tmpPath, filePath);
}

function hasAnosysHook(data) {
  const notifyList = data.notify;
  if (Array.isArray(notifyList)) {
    return notifyList.some(x => String(x).includes('anosys-codex'));
  }
  return false;
}

function updateCodexConfig(hookCmd, customPath) {
  const cmd = hookCmd || HOOK_COMMAND;
  const filePath = customPath || getConfigPath();
  const data = loadToml(filePath);

  let notifyList = data.notify || [];
  if (notifyList.includes(cmd)) {
    return false; // already present
  }

  notifyList = notifyList.filter(x => !String(x).includes('anosys-codex'));
  notifyList.push(cmd);
  data.notify = notifyList;

  writeTomlAtomic(filePath, data);
  return true;
}

function removeCodexConfig(customPath) {
  const filePath = customPath || getConfigPath();
  if (!fs.existsSync(filePath)) return false;

  const data = loadToml(filePath);
  let notifyList = data.notify || [];
  const prevLen = notifyList.length;

  notifyList = notifyList.filter(x => !String(x).includes('anosys-codex'));
  if (notifyList.length === prevLen) return false;

  data.notify = notifyList;
  writeTomlAtomic(filePath, data);
  return true;
}

function updateCodexEnv({ apiKey = '', redaction = false, endpointUrl = INGESTION_URL, customPath }) {
  const filePath = customPath || getEnvPath();
  const dir = path.dirname(filePath);
  fs.mkdirSync(dir, { recursive: true });

  const lines = [
    '# AnoSys Codex CLI configuration',
    `export ANOSYS_HOOK_ENDPOINT_URL="${endpointUrl}"`,
    `export REDACTION="${redaction ? 'true' : 'false'}"`
  ];
  if (apiKey) {
    lines.push(`export ANOSYS_HOOK_APIKEY="${apiKey}"`);
  }

  fs.writeFileSync(filePath, lines.join('\n') + '\n', { encoding: 'utf8', mode: 0o600 });
}

function removeCodexEnv(customPath) {
  const filePath = customPath || getEnvPath();
  if (fs.existsSync(filePath)) {
    try {
      fs.unlinkSync(filePath);
      return true;
    } catch (_) {
      return false;
    }
  }
  return false;
}

function validateApiKey(apiKey, keyType = 'codex') {
  return new Promise(resolve => {
    if (!apiKey) return resolve(false);
    const encoded = encodeURIComponent(apiKey);
    const url = `https://console.anosys.ai/api/resolveapikeys?apikey=${encoded}`;

    https.get(url, { headers: { 'User-Agent': 'anosys-codex-installer' }, timeout: 10000 }, res => {
      if (res.statusCode < 200 || res.statusCode >= 300) {
        return resolve(false);
      }
      let body = '';
      res.on('data', chunk => { body += chunk; });
      res.on('end', () => {
        try {
          const parsed = JSON.parse(body);
          resolve(Boolean(parsed && parsed.apiUrl));
        } catch (_) {
          resolve(false);
        }
      });
    }).on('error', () => resolve(false));
  });
}

module.exports = {
  HOOK_COMMAND,
  INGESTION_URL,
  getCodexHome,
  getConfigPath,
  getEnvPath,
  backup,
  loadToml,
  writeTomlAtomic,
  hasAnosysHook,
  updateCodexConfig,
  removeCodexConfig,
  updateCodexEnv,
  removeCodexEnv,
  validateApiKey
};
